### PR TITLE
TF-2628 [Part-2] Handle multiple attachment actions

### PR DESCRIPTION
**Pull request by tddang-linagora**
### **Issue**
#2628 

### **Improvement**

- Firefox

https://github.com/linagora/tmail-flutter/assets/160106668/42bba42c-c4f5-455e-b45b-1e5e322f53d7

- Safari

https://github.com/linagora/tmail-flutter/assets/160106668/d6635d4f-8c0f-4e0c-99cc-83b32d8045a2

- Chrome

https://github.com/linagora/tmail-flutter/assets/160106668/b7c03744-4f61-4d42-a38d-13c16d8797b0

### CODE REVIEW

File: lib/features/email/presentation/utils/email_utils.dart [hunk 1]
```diff
@@ -49,4 +51,32 @@ class EmailUtils {
...
+      return true;
+    }
+
+    if (state is ViewAttachmentForWebSuccess) {
+      return true;
+    }
+
+    if (state is ViewAttachmentForWebFailure) {
+      return true;
+    }
```
Comment (dab246) [inline lib/features/email/presentation/utils/email_utils.dart:78]: Should be use multiple conditions in the same `if`

File: lib/features/email/presentation/widgets/attachment_list/attachment_list_item_widget.dart [hunk 2]
```diff
@@ -8,86 +8,107 @@ import 'package:flutter/material.dart';
...
+                children: [
+                  isLoading
+                    ? const SizedBox(
+                      width: AttachmentListItemWidgetStyle.iconSize,
+                      height: AttachmentListItemWidgetStyle.iconSize,
+                      child: CircularProgressIndicator(strokeWidth: 2))
+                    : SvgPicture.asset(attachment.getIcon(_imagePaths),
+                      width: AttachmentListItemWidgetStyle.iconSize,
+                      height: AttachmentListItemWidgetStyle.iconSize,
+                      fit: BoxFit.fill
```
Comment (dab246) [inline lib/features/email/presentation/widgets/attachment_list/attachment_list_item_widget.dart:61]: indent

File: lib/features/email/presentation/widgets/attachment_list/attachment_list_item_widget.dart [hunk 3]
```diff
@@ -8,86 +8,107 @@ import 'package:flutter/material.dart';
...
                     )
-                  ]
-                )
+                  ),
+                  const SizedBox(width: AttachmentListItemWidgetStyle.space),
+                  TMailButtonWidget.fromIcon(
+                    icon: _imagePaths.icDownloadAttachment,
+                    backgroundColor: Colors.transparent,
+                    onTapActionCallback: isLoading ? null 
+                      : () => downloadAttachmentAction?.call(attachment)
```
Comment (dab246) [inline lib/features/email/presentation/widgets/attachment_list/attachment_list_item_widget.dart:104]:
```suggestion
                    onTapActionCallback: isLoading 
                      ? null 
                      : () => downloadAttachmentAction?.call(attachment)
```

File: lib/features/email/presentation/widgets/attachment_item_widget.dart [hunk 4]
```diff
@@ -25,71 +29,88 @@ class AttachmentItemWidget extends StatelessWidget {
...
+                    maxLines: 1,
+                    style: AttachmentItemWidgetStyle.sizeLabelTextStyle,
+                  ),
+                  TMailButtonWidget.fromIcon(
+                    icon: _imagePaths.icDownloadAttachment,
+                    backgroundColor: Colors.transparent,
+                    padding: const EdgeInsets.all(5),
+                    iconSize: AttachmentItemWidgetStyle.downloadIconSize,
+                    onTapActionCallback: isLoading ? null
+                      : () => downloadAttachmentAction?.call(attachment)
```
Comment (dab246) [inline lib/features/email/presentation/widgets/attachment_item_widget.dart:106]:
```suggestion
                    onTapActionCallback: isLoading 
                      ? null 
                      : () => downloadAttachmentAction?.call(attachment)
```

File: lib/features/email/domain/usecases/view_attachment_for_web_interactor.dart [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+                if (success is DownloadAttachmentForWebSuccess) {
+                  return right(ViewAttachmentForWebSuccess(
+                    success.taskId,
+                    success.attachment,
+                    success.bytes,
+                  ));
+                }
+
+                return right(success);
```
Comment (dab246) [inline lib/features/email/domain/usecases/view_attachment_for_web_interactor.dart:71]: Please use the keyword `Right/Left` is printed in capital letters for synchronization throughout the app, avoiding confusion for new users.

Comment (tddang-linagora) [reply lib/features/email/domain/usecases/view_attachment_for_web_interactor.dart:71]: I pushed this fixup at PR #2642 , please review there as this file wasn't created in this PR

File: lib/features/email/presentation/utils/email_utils.dart (hunk 1 above)
Comment (tddang-linagora) [reply lib/features/email/presentation/utils/email_utils.dart]: Done

File: lib/features/email/presentation/widgets/attachment_list/attachment_list_item_widget.dart (hunk 3 above)
Comment (tddang-linagora) [reply lib/features/email/presentation/widgets/attachment_list/attachment_list_item_widget.dart]: Done

File: lib/features/email/presentation/widgets/attachment_item_widget.dart (hunk 4 above)
Comment (tddang-linagora) [reply lib/features/email/presentation/widgets/attachment_item_widget.dart]: Done

File: lib/features/email/presentation/widgets/attachment_list/attachment_list_item_widget.dart (hunk 2 above)
Comment (tddang-linagora) [reply lib/features/email/presentation/widgets/attachment_list/attachment_list_item_widget.dart]: Done

File: lib/features/email/domain/state/download_attachment_for_web_state.dart [hunk 6]
```diff
@@ -56,9 +56,12 @@ class DownloadAttachmentForWebSuccess extends UIState {
 class DownloadAttachmentForWebFailure extends FeatureFailure {
 
   final DownloadTaskId? taskId;
+  final Attachment attachment;
```
Comment (hoangdat) [inline lib/features/email/domain/state/download_attachment_for_web_state.dart:61]: why we need `attachment` here?

File: lib/features/email/presentation/controller/single_email_controller.dart [hunk 7]
```diff
@@ -122,6 +122,7 @@ class SingleEmailController extends BaseController with AppLoaderMixin {
   final eventActions = <EventAction>[].obs;
   final emailLoadedViewState = Rx<Either<Failure, Success>>(Right(UIState.idle));
   final emailUnsubscribe = Rxn<EmailUnsubscribe>();
+  final attachmentsViewState = RxMap<Id, Object>();
```
Comment (hoangdat) [inline lib/features/email/presentation/controller/single_email_controller.dart:125]: why not `RxMap<Id, Either<Failure, Success>>>();`?

File: lib/features/email/presentation/utils/email_utils.dart [hunk 8]
```diff
@@ -49,4 +51,17 @@ class EmailUtils {
       return null;
     }
   }
+
+  static bool checkingIfAttachmentActionIsEnabled(Object? state) {
```
Comment (hoangdat) [inline lib/features/email/presentation/utils/email_utils.dart:55]: should use `Either`

File: lib/features/email/domain/state/download_attachment_for_web_state.dart (hunk 6 above)
Comment (tddang-linagora) [reply lib/features/email/domain/state/download_attachment_for_web_state.dart:61]: You're right. Its `blobId` is sufficient. Will change

Comment (tddang-linagora) [reply lib/features/email/domain/state/download_attachment_for_web_state.dart]: Done

File: lib/features/email/presentation/controller/single_email_controller.dart (hunk 7 above)
Comment (tddang-linagora) [reply lib/features/email/presentation/controller/single_email_controller.dart]: Done

File: lib/features/email/presentation/utils/email_utils.dart (hunk 8 above)
Comment (tddang-linagora) [reply lib/features/email/presentation/utils/email_utils.dart]: Done

File: lib/features/email/presentation/widgets/attachment_list/attachment_list_item_widget.dart [hunk 9]
```diff
@@ -31,66 +33,84 @@ class AttachmentListItemWidget extends StatelessWidget {
...
+                  isLoading
+                    ? const SizedBox(
+                        width: AttachmentListItemWidgetStyle.iconSize,
+                        height: AttachmentListItemWidgetStyle.iconSize,
+                        child: CircularProgressIndicator(strokeWidth: 2))
+                    : SvgPicture.asset(attachment.getIcon(_imagePaths),
+                        width: AttachmentListItemWidgetStyle.iconSize,
+                        height: AttachmentListItemWidgetStyle.iconSize,
+                        fit: BoxFit.fill
+                    ),
```
Comment (dab246) [inline lib/features/email/presentation/widgets/attachment_list/attachment_list_item_widget.dart:63]: indent

**Comment by dab246** [comment]
Please rebase

**Pull request merged**